File: 1036-rotting-oranges/rotting-oranges.py
class Solution:
    def orangesRotting(self, grid: List[List[int]]) -> int:
        levels = 0
        visited = [[0] * len(grid[0]) for _ in range(len(grid))]
        queue = []
        for r in range(len(grid)):
            for c in range(len(grid[0])):
                if grid[r][c] == 2:
                    queue.append((r,c))
        queue.append(None)
        while queue[0] != None:
            while queue[0] != None:
                ci,cj = queue.pop(0)
                for di,dj in [(1,0),(0,1),(-1,0),(0,-1)]:
                    ni,nj = di + ci, dj + cj
                    if ni >= 0 and ni < len(grid) and nj >= 0 and nj < len(grid[0]) and grid[ni][nj] == 1 and visited[ni][nj] == 0:
                        visited[ni][nj] = 1
                        queue.append((ni,nj))
                        grid[ni][nj] = 2
            queue.append(queue.pop(0))
            levels+=1
        for r in range(len(grid)):
            for c in range(len(grid[0])):
                if grid[r][c] == 1:
                    return -1
        if levels <= 0:
            return 0
        return levels - 1

File: 1036-rotting-oranges/rotting-oranges.java
class Solution {
    public int orangesRotting(int[][] grid) {
        Queue<Pair> queue = new LinkedList<>();
        int [] xNext = new int[]{1,-1,0,0};
        int [] yNext = new int[]{0,0,1,-1};
        int fresh = 0;
        int rotten = 0;
        for(int i = 0;i<grid.length;i++){
            for(int j = 0;j<grid[0].length;j++){
                if(grid[i][j] == 2){
                    queue.add(new Pair(i,j));
                    rotten +=1;
                }
                else if(grid[i][j] == 1){
                    fresh+=1;
                }
            }
        }
        int result = 0;
        queue.add(null);
        while(queue.size() > 1){
            while(queue.peek() != null){
                Pair curr = queue.poll();
                for(int i = 0;i<4;i++){
                    int nextx = curr.x + xNext[i];
                    int nexty = curr.y + yNext[i];
                    if(nextx >= 0 && nextx < grid.length && nexty >= 0 && nexty < grid[0].length && grid[nextx][nexty] == 1){
                        queue.add(new Pair(nextx,nexty));
                        grid[nextx][nexty] = 2;
                        fresh-=1;
                    }
                }
            }
            queue.poll();
            queue.add(null);
            result+=1;
        }
        if(fresh > 0){
            return -1;
        }
        if(rotten == 0){
            return 0;
        }
        return result-1;
    }
}
class Pair{
    int x;
    int y;
    public Pair(int x, int y){
        this.x = x;
        this.y = y;
    }
}